### DISCUSSION TITLE
cellPlugin with controls = { type 'custom'}: how does one get react data in?

### DISCUSSION THREAD

**Discussion by zehawki** [react-page/react-page#902] [Q&A] [answered]
So I've gotten custom `cellPlugin` working on my side with` { type: 'autoform' }`, and it seems to be clear that the data is going from the `BottomToolbar` control (form) into the `Renderer`. So far so good.

Now I'm stuck at `{ type 'custom'}`:

Using `createInitialData`, I've managed to create dummy data that will populate the `customCell` when inserted. So far so good again.

Now for the life of me I cannot figure out how to get real React data from props, store etc into this component.

Perhaps a simpler question is: how does one get access to `props` inside the custom plugin?

**Comment by zehawki** [comment]
I see this #443, but I'm no wiser :-D

**Reply by zehawki** [reply]
@macrozone whenever you are able to give me some clues on this... I've look thru all the previous issue and code examples etc, but not able to figure it out.

**Comment by macrozone** [comment, marked as answer]
the Bottom Toolbar renders a component that updates the `data` on the cell.
The `type: "autoform"` in particular renders a form automatically based on a JSONSchema (https://json-schema.org/), so you don't have to worry about writing any `<input />` or validation logic at all.

https://react-page.github.io/docs/#/custom-cell-plugins shows an example where it just takes a title. So the bottom toolbar would render an input for that title and changing it would set `data: {title: "whatever"}` to the props that get passed to your plugin's Component.

It also shows how your Component (Renderer) receives this data. (prop `data`).

Another example is the Twitter  Feed plugin on the demo page: https://react-page.github.io/ (scroll down), the demo also shows the true source code of that plugin next to it.

Initially, if a cell with this custom plugin is added, the data is null, unless you specify `createInitialData` in which case it is whatever this function returns. This is usefull if you want to pre-fill data for newly added cells (but its also possible to define defaults to the json-schema).

**Reply by zehawki** [reply]
Right, everything you've mentioned here is regarding `{ type: 'autoform' }`, which is clear.

But I was asking about `{ type 'custom'}` where I want to show some arbitrary (existing) React component on the page, which is not related to user input via a form. This React component has its own props which need to be passed through, so how is that to be done?

**Reply by macrozone** [reply]
don't confuse controls with the Renderer.

controls: what to show in the BottomToolbar.
Renderer: what to show on the Page

if you the Component you want to add does not require any input at all, don't specify `controls`. Only use `type: "custom"` if you really need a very special sophisticated form in the bottom toolbar.

What Renderer is doing, is up to you. You can show an image gallery, some product teaser, a 3d object, whatever you want. You can use any arbitrary React Component.

If this arbitrary component needs data e.g. from an api, it will probably use some hook (e.g. apollo's useQuery, or useSWR, or load data from redux store, etc.). 

E.g. take this example: https://react-page.github.io/examples/reactadmin#/posts/post1

There you see a list or Products as Cards. If you click on it, you will see that you can select multiple products.
Now look at how one of these Product  teasers is defined: https://github.com/react-page/react-page/blob/master/examples/pages/examples/reactadmin.tsx#L145

Its just an arbitrary Component that takes a productId and renders title, intro and image of that product. Later down here: https://github.com/react-page/react-page/blob/master/examples/pages/examples/reactadmin.tsx#L177 is the definition of the CellPlugin, its Renderer takes a list or productIds and renders a list of ProductTeasers. 
Which products to render can be defined by the user that uses the editor via the BottomToolbar, but how the ProductTeaser loads a product from a given productId is none of our business and highly depends on the rest of your application.
If a user adds multiple of these Product teaser lists to the page, they can have different productIds selected.

This is how you usually do a react application, you normally don't pass everything down from the root component, but you use  data loading libraries like @apollo/client, firebase or local state management like redux and easy peasy, etc. All of these libraries use some kind of React-Context.

To answer https://github.com/react-page/react-page/issues/443: there is no need to pass props from editor down to all the plugins, react already provides the tools you need.

**Reply by zehawki** [reply]
Marco, thank you for the very detailed reply.

What you are saying is very clear and I've already looked at all these resources :-)

> controls: what to show in the BottomToolbar.
> Renderer: what to show on the Page

Yup, I figured that.

> if you the Component you want to add does not require any input at all, don't specify controls

Yes, I'm not using control

> What Renderer is doing, is up to you. You can show an image gallery, some product teaser, a 3d object, whatever you want. You can use any arbitrary React Component.

Also clear, this is fine and obvious. Thats not the point.

> Which products to render can be defined by the user that uses the editor via the BottomToolbar,

This is the point. This example is also about type: 'autoform', where the INPUT comes from the user. The product ID is flowing though from the user via data props in to the component and then the component knows what to do with it. (I dont care at all what the component does next - obviously the component knows what to do once props have been received,)

What I'm asking is: _how do I handle the props going to the component WITHOUT user input_. In this same example that you have provided, say props productId has to come from somewhere else (eg like I said above, from redux store, or elsewhere, doesnt matter), how is that to be sent to the component. Its not going to be based on user selection.

**Reply by macrozone** [reply]
in this case, you would have:

```
const myPlugin = {
   // id, title and stuff...
  Renderer: () => {
     // in case of redux:
     const productId = useSelector(state => state.to.productId)
     return <ProductTeaser productId={productId} />
   }
}
```

Edit: this is just react or component driven development in general. you create new components by composition. So naturally you can also do:

```
const MyProductTeaserThatKnowsHowToLoadData = () => {
     // in case of redux:
     const productId = useSelector(state => state.to.productId)
     return <ProductTeaser productId={productId} />
}
const myPlugin = {
   // id, title and stuff...
  Renderer: MyProductTeaserThatKnowsHowToLoadData
}
```

**Reply by zehawki** [reply]
You sir are a lifesaver. `useSelector`!!! But of course. I've been mucking around with `connect() ` on and on, without getting anywhere.

**Reply by macrozone** [reply]
connect works as well of course:

```
const mapStateToProps = state => ({productId: state.to.productId})
const MyProductTeaserThatKnowsHowToLoadData = connect(mapStateToProps)({productId}) => {
     return <ProductTeaser productId={productId} />
}
const myPlugin = {
   // id, title and stuff...
  Renderer: MyProductTeaserThatKnowsHowToLoadData
}
```

but hooks are in general much simpler and easier to use than the older connect/"HOC"-approach ("higher order components")

**Reply by zehawki** [reply]
Yeah been writing new modules with hooks, and migrating my older code one by one to functional. Still end up thinking class / HOC first.
